packages/core/src/extension/extension-service.ts

```diff
@@ -0,0 +1,58 @@
+import { noop } from 'lodash';
+
+import { IframeEventListener, IframeEventPoster } from '../common/iframe';
+
+import {
+    ExtensionListenEventMap,
+    ExtensionListenEventType,
+    ExtensionPostEvent,
+    ExtensionPostEventType,
+} from './extension-client';
+
+export default class ExtensionService {
+    private _extensionId?: string;
+
+    constructor(
+        private _eventListener: IframeEventListener<ExtensionListenEventMap>,
+        private _eventPoster: IframeEventPoster<ExtensionPostEvent>,
+    ) {
+        this._eventPoster.setTarget(window.parent);
+    }
+
+    initialize(extensionId: string) {
+        if (!extensionId) {
+            throw new Error('Extension Id not found.');
+        }
+
+        this._extensionId = extensionId;
+
+        this._eventListener.listen();
+    }
+
+    post(event: ExtensionPostEvent): void {
+        if (!this._extensionId) {
+            return;
+        }
+
+        if (!Object.values(ExtensionPostEventType).includes(event.type)) {
+            throw new Error(`${event.type} is not supported.`);
+        }
+
+        const payload = {
+            ...event.payload,
+            extensionId: this._extensionId,
+        };
+
+        this._eventPoster.post({ ...event, payload });
+    }
+
+    addListener(eventType: ExtensionListenEventType, callback: () => void = noop): () => void {
```

Comment:
It seems we need to standardise the naming. There is `Command`, `ListenEvent`, `PostEvent`, `OriginEvent`, `BroadcastEvent` etc... I'm not exactly sure what they are and whether they need to be categorised differently.